### ISSUE TITLE
AOS lower ap compared with other metric

### ISSUE THREAD

**Issue by buaaJeremyduan** [traveller59/second.pytorch#238]
Car AP: 0.7, 0.7, 0.7
bbox: 60.53, 51.58, 50.10
bev: 87.79, 77.43, 71.79
3d: 22.41, 22.55, 21.09
aos: 2.70, 11.64, 12.51
This is the result of my first evaluation result when training car.lite.config, Is that normal? I notice the ap of aos is much lower than bev and bbox. Does this mean that it is hard to learn the orientation of the box? Thanks.

**Comment by micjabbour** [comment]
I have been tracking down a similar issue in AOS evaluation results for a while. When visualizing the results, one can clearly see that the network gets the direction wrong most of the time.

### Summary:
This seems to be caused by an inconsistency in interpreting the `dir_cls_preds` values predicted by the `RPN` when calculating the loss function and when performing a prediction. The issue can be fixed by changing the `direction_limit_offset` parameter in the config file from `1` to `0`. Please note that this parameter does not currently affect the training process in any way. You don't need to re-train your model, just re-run evaluation after updating the configuration parameter.

### Details:
The problematic changes are introduced in the commit [`e42e4a0`](https://github.com/traveller59/second.pytorch/commit/e42e4a0e17262ab7d180ee96a0a36427f2c20a44), when many configuration parameters have been added to the direction classifier. The output from the direction classifier seems to be read in a way that conflicts with the way it is trained. The following explanation assumes that `num_direction_bins` is set to `2` and `direction_offset` is set to `0` (which is the case in [many](https://github.com/traveller59/second.pytorch/tree/e42e4a0e17262ab7d180ee96a0a36427f2c20a44/second/configs/pointpillars) [of](https://github.com/traveller59/second.pytorch/blob/e42e4a0e17262ab7d180ee96a0a36427f2c20a44/second/configs/all.fhd.config) [the](https://github.com/traveller59/second.pytorch/blob/e42e4a0e17262ab7d180ee96a0a36427f2c20a44/second/configs/car.fhd.config) [provided](https://github.com/traveller59/second.pytorch/blob/e42e4a0e17262ab7d180ee96a0a36427f2c20a44/second/configs/car.fhd.onestage.config) [configuration](https://github.com/traveller59/second.pytorch/blob/e42e4a0e17262ab7d180ee96a0a36427f2c20a44/second/configs/car.lite.config) [files](https://github.com/traveller59/second.pytorch/blob/e42e4a0e17262ab7d180ee96a0a36427f2c20a44/second/configs/people.fhd.config)):

- When calculating the direction loss, the targets are set such that the class is `0` for yaw values between `0` and `pi`, and `1` for yaw values between `pi` and `2 * pi` (see the function [`get_direction_target`](https://github.com/traveller59/second.pytorch/blob/e42e4a0e17262ab7d180ee96a0a36427f2c20a44/second/pytorch/models/voxelnet.py#L814))

- [Upon prediction](https://github.com/traveller59/second.pytorch/blob/e42e4a0e17262ab7d180ee96a0a36427f2c20a44/second/pytorch/models/voxelnet.py#L601), `limit_period` is called with `period=1` (the wrong value of `direction_limit_offset`), which causes `limit_period`'s output to be always between `[-pi, 0[`. Afterwards, an additional period of `pi` is added iff the predicted direction class is `1`. This means that we get predicted yaw angles between `0` and `pi` when the predicted direction class is `1`, and angles between `-pi` and `0` when the predicted direction class `0` (the exact opposite of the way the direction targets are set). Setting the period to `0` in the call to `limit_period` fixes the issue.

Hope this helps...

**Comment by buaaJeremyduan** [comment]
Thanks for your help. I will try it out.

**Comment by buaaJeremyduan** [comment]
Car AP(Average Precision)@0.70, 0.70, 0.70:
bbox AP:97.18, 89.44, 88.62
bev  AP:89.89, 88.09, 86.89
3d   AP:87.84, 78.34, 77.26
aos  AP:97.08, 89.09, 88.06
Car AP(Average Precision)@0.70, 0.50, 0.50:
bbox AP:97.18, 89.44, 88.62
bev  AP:97.44, 89.74, 89.20
3d   AP:97.36, 89.68, 89.07
aos  AP:97.08, 89.09, 88.06

Evaluation coco
Car coco AP@0.50:0.05:0.95:
bbox AP:76.24, 69.60, 68.55
bev  AP:72.25, 67.56, 65.65
3d   AP:62.44, 57.50, 56.44
aos  AP:76.17, 69.36, 68.14

I follow your instruction and I got the result above.

**Comment by BB88Lee** [comment]
which config did you use? @buaaJeremyduan 

 with car.lite.config got result as follow:

Car AP(Average Precision)@0.70, 0.70, 0.70:
bbox AP:90.62, 88.92, 87.30
bev  AP:89.84, 86.03, 79.59
3d   AP:86.81, 76.87, 74.67
aos  AP:90.53, 88.47, 86.55
Car AP(Average Precision)@0.70, 0.50, 0.50:
bbox AP:90.62, 88.92, 87.30
bev  AP:90.70, 89.68, 88.87
3d   AP:90.70, 89.61, 88.63
aos  AP:90.53, 88.47, 86.55

Evaluation coco 
Car coco AP@0.50:0.05:0.95:
bbox AP:72.34, 68.60, 66.33
bev  AP:69.38, 65.63, 63.48
3d   AP:60.05, 55.66, 53.64
aos  AP:72.27, 68.28, 65.81

**Comment by buaaJeremyduan** [comment]
car.fhd.config
